File: src/compiler/lexer/lexer.zig
const std = @import("std");

pub const TokenKind = enum {
    undef,

    integer_number_literal,       // e.g.: 123, 0x10, 0b0110
    floating_number_literal,      // e.g.: 1.23
    string_literal,               // e.g.: "bruh"
    character_literal,            // e.g.: 'a', '\r'
    escaped_character_literal,    // e.g.: 'a', '\r'
    identifier,                   // e.g.: variable_name

    namespace_keyword,            // namespace
    import_keyword,               // import
    from_keyword,                 // from
    let_keyword,                  // let
    const_keyword,                // const
    func_keyword,                 // func
    struct_keyword,               // struct
    extends_keyword,              // extends
    packet_keyword,               // packet
    enum_keyword,                 // enum

    switch_keyword,               // switch
    match_keyword,                // match

    if_keyword,                   // if
    elif_keyword,                 // elif
    else_keyword,                 // else
    while_keyword,                // while
    for_keyword,                  // for
    do_keyword,                   // do
    in_keyword,                   // in
    break_keyword,                // break

    as_keyword,                   // as
    return_keyword,               // return

    null_keyword,                 // null
    true_keyword,                 // true
    false_keyword,                // false

    left_parenthesis_char,        // (
    right_parenthesis_char,       // )

    left_brace_char,              // {
    right_brace_char,             // }

    left_square_bracket_char,     // [
    right_square_bracket_char,    // ]

    left_angle_char,              // <
    right_angle_char,             // >

    escaped_left_brace_char,      // \{

    plus_char,                    // +
    minus_char,                   // -
    star_char,                    // *
    slash_char,                   // /
    reverse_slash_char,           // \
    percent_char,                 // %
    equals_char,                  // =
    ampersand_char,               // &
    question_char,                // ?
    bang_char,                    // !

    at_sign_char,                 // @

    right_arrow_operator,         // =>

    equal_operator,               // ==
    not_equal_operator,           // !=
    less_equal_operator,          // <=
    greater_equal_operator,       // >=

    and_operator,                 // and
    or_operator,                  // or

    power_operator,               // **

    add_assign_operator,          // +=
    sub_assign_operator,          // -=
    mul_assign_operator,          // *=
    div_assign_operator,          // /=
    mod_assign_operator,          // %=
    pow_assign_operator,          // **=

    increment_operator,           // ++
    decrement_operator,           // --

    range_operator,               // ..

    single_quote_char,            // '
    double_quote_char,            // "

    comma_char,                   // ,
    dot_char,                     // .

    eof,                          // \EOF
    line_feed,                    // \n
};
pub const Token = struct {
    kind: TokenKind,
    value: ?[]const u8 = null,
};

const dictionary = [_]struct { []const u8, TokenKind } {
    // keywords
    .{"namespace", TokenKind.namespace_keyword },
    .{"import", TokenKind.import_keyword },
    .{"from", TokenKind.from_keyword },

    .{"let", TokenKind.let_keyword },
    .{"const", TokenKind.const_keyword },
    .{"func", TokenKind.func_keyword },
    .{"struct", TokenKind.struct_keyword },
    .{"extends", TokenKind.extends_keyword },
    .{"packet", TokenKind.packet_keyword },
    .{"enum", TokenKind.enum_keyword },

    .{"switch", TokenKind.switch_keyword },
    .{"match", TokenKind.match_keyword },
    .{"if", TokenKind.if_keyword },
    .{"elif", TokenKind.elif_keyword },
    .{"else", TokenKind.else_keyword },

    .{"while", TokenKind.while_keyword },
    .{"for", TokenKind.for_keyword },
    .{"do", TokenKind.do_keyword },
    .{"in", TokenKind.in_keyword },
    .{"break", TokenKind.break_keyword },

    .{"return", TokenKind.return_keyword },

    .{"as", TokenKind.as_keyword },

    // values
    .{"null", TokenKind.null_keyword },
    .{"true", TokenKind.true_keyword },
    .{"false", TokenKind.false_keyword },

    // operators
    .{"and", TokenKind.and_operator },
    .{"or", TokenKind.or_operator },
};
const line_juctions = .{
    .only_left = .{
        TokenKind.right_brace_char,
        TokenKind.right_parenthesis_char,
    },
    .only_right = .{
        TokenKind.left_brace_char,
        TokenKind.left_square_bracket_char,
        TokenKind.escaped_left_brace_char,
        TokenKind.comma_char,
    },
    .both = .{
        TokenKind.dot_char,
        TokenKind.range_operator,

        TokenKind.plus_char, TokenKind.minus_char, TokenKind.power_operator,
        TokenKind.star_char, TokenKind.percent_char, TokenKind.equals_char,

        TokenKind.right_arrow_operator, TokenKind.less_equal_operator,
        TokenKind.equal_operator, TokenKind.not_equal_operator,
        TokenKind.left_angle_char,  TokenKind.right_angle_char,
        TokenKind.less_equal_operator, TokenKind.greater_equal_operator,

        TokenKind.add_assign_operator, TokenKind.mul_assign_operator,
        TokenKind.sub_assign_operator, TokenKind.div_assign_operator,
        TokenKind.mod_assign_operator, TokenKind.pow_assign_operator,
    },
};
const lang_symbols: []const u8  = "=+-*/!@$%|:;.?<>";
const lang_numerics: []const u8 = "0123456789abcdef";

pub fn lex_slice(allocator: std.mem.Allocator, source: []const u8) ![]const Token {
    var token_list: std.ArrayList(Token) = .init(allocator);
    defer token_list.deinit();

    var i: usize = 0;
    while (i < source.len) : (i += 1) {

        const c = source[i];
        const c2 = if (i+1 < source.len) source[i+1] else 0;


        // Check skipable
        if (c == ' ' or c == '\t') { continue; }

        // Check if it's a multiline character
        if (is_language_symbol(c) and is_language_symbol(c2)) {

            const cc: []const u8 = &.{ c, c2 };

            const tkind =
                 if (std.mem.eql(u8, cc, "=>")) TokenKind.right_arrow_operator
            else if (std.mem.eql(u8, cc, "==")) TokenKind.equal_operator
            else if (std.mem.eql(u8, cc, "!=")) TokenKind.not_equal_operator
            else if (std.mem.eql(u8, cc, "<=")) TokenKind.less_equal_operator
            else if (std.mem.eql(u8, cc, ">=")) TokenKind.greater_equal_operator
            else if (std.mem.eql(u8, cc, "**")) TokenKind.power_operator
            else if (std.mem.eql(u8, cc, "+=")) TokenKind.add_assign_operator
            else if (std.mem.eql(u8, cc, "-=")) TokenKind.sub_assign_operator
            else if (std.mem.eql(u8, cc, "*=")) TokenKind.mul_assign_operator
            else if (std.mem.eql(u8, cc, "/=")) TokenKind.div_assign_operator
            else if (std.mem.eql(u8, cc, "%=")) TokenKind.mod_assign_operator
            else TokenKind.undef;

            if (tkind != TokenKind.undef) {
                try token_list.append(.{ .kind = tkind });
                i += 1;
                continue;
            }

        }

        // Check if line feed
        if (c == '\n' or c == '\r') {
            try token_list.append(.{ .kind = .line_feed  });
            if (c == '\r' and c2 == '\n') i+=1;
            continue;
        }

        // Check if single character
        var match: bool = true;
        switch (c) {
            '(' => try token_list.append(.{ .kind = .left_parenthesis_char }),
            ')' => try token_list.append(.{ .kind = .right_parenthesis_char }),
            '{' => try token_list.append(.{ .kind = .left_brace_char }),
            '}' => try token_list.append(.{ .kind = .right_brace_char }),
            '[' => try token_list.append(.{ .kind = .left_square_bracket_char }),
            ']' => try token_list.append(.{ .kind = .right_square_bracket_char }),
            
            '<' => try token_list.append(.{ .kind = .left_angle_char }),
            '>' => try token_list.append(.{ .kind = .right_angle_char }),
            
            '+' => try token_list.append(.{ .kind = .plus_char }),
            '-' => try token_list.append(.{ .kind = .minus_char }),
            '*' => try token_list.append(.{ .kind = .star_char }),
            '/' => try token_list.append(.{ .kind = .slash_char }),
            '%' => try token_list.append(.{ .kind = .percent_char }),
            '=' => try token_list.append(.{ .kind = .equals_char }),

            '&' => try token_list.append(.{ .kind = .ampersand_char }),
            '?' => try token_list.append(.{ .kind = .question_char }),
            '!' => try token_list.append(.{ .kind = .bang_char }),
            '@' => try token_list.append(.{ .kind = .at_sign_char }),

            ',' => try token_list.append(.{ .kind = .comma_char }),
            '.' => try token_list.append(.{ .kind = .dot_char }),

            else => match = false
        }
        if (match) continue;

        // Build number
        if (std.ascii.isDigit(c)) {

            var j: usize = i;
            var base: usize = 10;
            var is_floating: bool = false;
            if (c == 0) {
                j += 1;
                switch (c2) {
                    'x' => base = 16,
                    'b' => base = 2,
                    'o' => base = 8,
                    else => j -= 1
                }
            }

            while (j+1 < source.len) : (j+=1) {
                const c3 = source[j+1];
                if (c3 == '.') is_floating = true;
                if (c3 != '_' and c3 != '.' and !is_digit(c3, base)) break;
            }

            const v = try allocator.dupe(u8, source[i..j+1]);
            errdefer allocator.free(v);

            try token_list.append(.{ 
                .kind = if (is_floating) .floating_number_literal else .integer_number_literal,
                .value = v,
            });

            i = j;
            continue;

        }

        // Check word
        if (is_valid_identifier_starter(c)) {

            var j = i;
            while (j+1 < source.len) : (j += 1) {
                if (!is_valid_identifier(source[j+1])) break;
            }

            const nameslice = try allocator.dupe(u8, source[i..j+1]);
            errdefer allocator.free(nameslice);

            const is_in_dic, const kind = b: {
                for (dictionary) |k| {
                    const str = k.@"0";
                    const tkn = k.@"1";

                    if (std.mem.eql(u8, str, nameslice)) break :b .{ true, tkn };
                }
                break :b .{ false, undefined };
            };

            if (is_in_dic) {
                try token_list.append(.{ .kind = kind });
            }
            else {
                try token_list.append(.{
                    .kind = .identifier,
                    .value = nameslice
                });
            }

            i = j;
            continue;

        }

        // Check string
        // TODO implement escaped characters and interpolation
        if (c == '"') {
            try token_list.append(.{ .kind = .double_quote_char });

            var j: usize = i;
            while (j+1 > source.len) : (j += 1) {
                if (source[j+1] == '"') {
                    try token_list.append(.{
                        .kind = .string_literal,
                        .value = source[i..j]
                    });
                    try token_list.append(.{ .kind = .double_quote_char });
                    break;
                }
            }

            i = j;
        }
        if (c == '\'') {

        }

        // Ignore comments
        if (c == '#') {
            if (source.len > i + 3 and std.mem.eql(u8, source[i..i+3], "###")) {
                i += 3;
                while (source.len > i+3 and !std.mem.eql(u8, source[i..i+3], "###")) : (i += 1) {}
                i += 3;
            } else {
                i += 1;
                while (source.len > i+1 and source[i] != '\n') : (i += 1) {}
            }
        }

    }

    return token_list.toOwnedSlice();
}
pub fn reduce_line_feeds(allocator: std.mem.Allocator, tokens: []const Token) ![]const Token {

    var tkn_list = std.ArrayList(Token).fromOwnedSlice(allocator, @constCast(tokens));

    for (tkn_list.items, 0..) |t, i| {
        if (t.kind == .line_feed) {
            while (tkn_list.items.len > i+1 and tkn_list.items[i+1].kind == .line_feed) {
                _ = tkn_list.orderedRemove(i+1);
            }
        }
    }

    return tkn_list.toOwnedSlice();

}
pub fn free_tokens(allocator: std.mem.Allocator, tokens: []const Token) void {
    for (tokens) |i| if (i.value) |v| allocator.free(v);
    allocator.free(tokens);
}

inline fn is_digit(char: u8, base: usize) bool {
    return std.mem.containsAtLeastScalar(u8, lang_numerics[0..base], 1, std.ascii.toLower(char));
}
inline fn is_language_symbol(char: u8) bool {
    return std.mem.containsAtLeastScalar(u8, lang_symbols, 1, char);
}
inline fn is_valid_identifier_starter(char: u8) bool {
    return char == '_' or std.ascii.isAlphabetic(char);
}
inline fn is_valid_identifier(char: u8) bool {
    return char == '_' or std.ascii.isAlphanumeric(char);
}

const expect = std.testing.expect;
const test_allocator = std.testing.allocator;
test "lexer general" {

    var arena = std.heap.ArenaAllocator.init(test_allocator);
    const allocator = arena.allocator();

    const sample = 
    \\@public const f64 PI = 3.1415
    \\@public const f32 E = 2.71828
    \\
    \\# This shouold not be tokenized!
    \\
    \\@public func sin(type T, T rad) T {
    \\      call("Hello, World!")
    \\}
    \\
    ;

    const tkns = try reduce_line_feeds(allocator, try lex_slice(allocator, sample));

    try expect(tkns[0].kind == .at_sign_char);
    try expect(tkns[1].kind == .identifier and std.mem.eql(u8, tkns[1].value.?, "public"));
    try expect(tkns[2].kind == .const_keyword);
    try expect(tkns[3].kind == .identifier and std.mem.eql(u8, tkns[3].value.?, "f64"));
    try expect(tkns[4].kind == .identifier and std.mem.eql(u8, tkns[4].value.?, "PI"));
    try expect(tkns[5].kind == .equals_char);
    try expect(tkns[6].kind == .floating_number_literal and std.mem.eql(u8, tkns[6].value.?, "3.1415"));
    try expect(tkns[7].kind == .line_feed);

    try expect(tkns[8].kind == .at_sign_char);
    try expect(tkns[9].kind == .identifier and std.mem.eql(u8, tkns[9].value.?, "public"));
    try expect(tkns[10].kind == .const_keyword);
    try expect(tkns[11].kind == .identifier and std.mem.eql(u8, tkns[11].value.?, "f32"));
    try expect(tkns[12].kind == .identifier and std.mem.eql(u8, tkns[12].value.?, "E"));
    try expect(tkns[13].kind == .equals_char);
    try expect(tkns[14].kind == .floating_number_literal and std.mem.eql(u8, tkns[14].value.?, "2.71828"));
    try expect(tkns[15].kind == .line_feed);

    try expect(tkns[16].kind == .at_sign_char);
    try expect(tkns[17].kind == .identifier and std.mem.eql(u8, tkns[17].value.?, "public"));
    try expect(tkns[18].kind == .func_keyword);
    try expect(tkns[19].kind == .identifier and std.mem.eql(u8, tkns[19].value.?, "sin"));
    try expect(tkns[20].kind == .left_parenthesis_char);
    try expect(tkns[21].kind == .identifier and std.mem.eql(u8, tkns[21].value.?, "type"));
    try expect(tkns[22].kind == .identifier and std.mem.eql(u8, tkns[22].value.?, "T"));
    try expect(tkns[23].kind == .comma_char);
    try expect(tkns[24].kind == .identifier and std.mem.eql(u8, tkns[24].value.?, "T"));
    try expect(tkns[25].kind == .identifier and std.mem.eql(u8, tkns[25].value.?, "rad"));
    try expect(tkns[26].kind == .right_parenthesis_char);
    try expect(tkns[27].kind == .identifier and std.mem.eql(u8, tkns[27].value.?, "T"));

    try expect(tkns[28].kind == .left_brace_char);
    try expect(tkns[29].kind == .line_feed);

    try expect(tkns[30].kind == .identifier and std.mem.eql(u8, tkns[30].value.?, "call"));
    try expect(tkns[31].kind == .left_parenthesis_char);
    try expect(tkns[32].kind == .double_quote_char);
    try expect(tkns[33].kind == .string_literal and std.mem.eql(u8, tkns[33].value.?, "Hello, World!"));
    try expect(tkns[34].kind == .double_quote_char);
    try expect(tkns[35].kind == .line_feed);
    try expect(tkns[36].kind == .right_parenthesis_char);

    try expect(tkns[37].kind == .right_brace_char);

    arena.deinit();

}

File: Abstract/Abstract.CodeProcess/Lexer.cs
using System.Diagnostics;
using Abstract.CodeProcess.Core.Language;

namespace Abstract.CodeProcess;

public class Lexer
{

    private readonly Dictionary<string, TokenType[]> lineJunctions = new()
    {
        { "justLeft",  [
            TokenType.LeftBracketChar,
            TokenType.RightParenthesisChar
        ] },
        { "justRight", [
            TokenType.LeftPerenthesisChar,
            TokenType.AtSiginChar,
        ] },
        { "bothSides", [
            TokenType.DotChar,
            TokenType.CrossChar, TokenType.MinusChar, TokenType.PowerOperator,
            TokenType.StarChar, TokenType.StarChar, TokenType.PercentChar,
            TokenType.EqualsChar,
            
            TokenType.LeftBracketChar, //TokenType.RightBracketChar,

            TokenType.RightArrowOperator, TokenType.LessEqualsOperator,
            TokenType.EqualOperator, TokenType.UnEqualOperator,
            TokenType.LeftAngleChar, TokenType.RightAngleChar,
            TokenType.LessEqualsOperator, TokenType.GreatEqualsOperator,

            TokenType.AddAssigin, TokenType.MulAssigin,
            TokenType.SubAssigin, TokenType.DivAssigin
        ] },
    };

    private Dictionary<string, TokenType> _keyword2TokenMap = new()
    {
        // keywords
        { "from", TokenType.FromKeyword },
        { "import", TokenType.ImportKeyword },

        { "let", TokenType.LetKeyword },
        { "const", TokenType.ConstKeyword },
        { "func", TokenType.FuncKeyword },
        { "struct", TokenType.StructKeyword },
        { "extends", TokenType.ExtendsKeyword },
        { "packet", TokenType.PacketKeyword },
        { "typedef", TokenType.TypedefKeyword },

        { "switch", TokenType.SwitchKeyword },
        { "match", TokenType.MatchKeyword },
        { "if", TokenType.IfKeyword },
        { "elif", TokenType.ElifKeyword },
        { "else", TokenType.ElseKeyword },

        { "while", TokenType.WhileKeyword },
        { "for", TokenType.ForKeyword },
        { "do", TokenType.DoKeyword },
        { "in", TokenType.InKeyword },
        { "break", TokenType.BreakKeyword },

        { "return", TokenType.ReturnKeyword },

        { "as", TokenType.AsKeyword },
        { "new", TokenType.NewKeyword },

        // values
        { "null", TokenType.NullKeyword },
        { "true", TokenType.TrueKeyword },
        { "false", TokenType.FalseKeyword },

        // operators
        { "and", TokenType.AndOperator },
        { "or", TokenType.OrOperator },

        // types
        { "void", TokenType.TypeKeyword },
        { "noreturn", TokenType.TypeKeyword },
        { "type", TokenType.TypeKeyword },

        { "byte", TokenType.TypeKeyword },
        { "f32", TokenType.TypeKeyword },   { "float", TokenType.TypeKeyword },
        { "f64", TokenType.TypeKeyword },   { "double", TokenType.TypeKeyword },

        { "bool", TokenType.TypeKeyword },
        { "char", TokenType.TypeKeyword },
        { "string", TokenType.TypeKeyword },

    };

    private Token Tokenize(TokenType type, (ReadOnlyMemory<char> val, uint line, uint start, uint end) range)
        => new() { type = type, value = range.val, line = range.line, start = range.start,  end = range.end};
    
    public Token[] Lex(string source)
    {
        List<Token> tokens = [];
        
        LexRange(source, tokens);
        VerifyEndOfStatements(tokens);

        return [.. tokens];
    }
    
    private void LexRange(string source, List<Token> tokens)
    {
        SourceWindow src = new(source);

        while (!src.IsEof())
        {
            char c = src.Next();
            
            // Check if it's skipable
            if (c == ' ' | c == '\r' | c == '\t')
            {
                src.Discard();
                continue;
            }


            switch (c)
            {
                // Check single characters
                case '\n': tokens.Add(Tokenize(TokenType.LineFeedChar, src.GetSlice())); break;
                
                case '(': tokens.Add(Tokenize(TokenType.LeftPerenthesisChar, src.GetSlice())); break;
                case ')': tokens.Add(Tokenize(TokenType.RightParenthesisChar, src.GetSlice())); break;
                case '{': tokens.Add(Tokenize(TokenType.LeftBracketChar, src.GetSlice())); break;
                case '}': tokens.Add(Tokenize(TokenType.RightBracketChar, src.GetSlice())); break;
                case '[': tokens.Add(Tokenize(TokenType.LeftSquareBracketChar, src.GetSlice())); break;
                case ']': tokens.Add(Tokenize(TokenType.RightSquareBracketChar, src.GetSlice())); break;
                case '&': tokens.Add(Tokenize(TokenType.AmpersandChar,src.GetSlice())); break;
                case '?': tokens.Add(Tokenize(TokenType.QuestionChar, src.GetSlice())); break;
                case '!': tokens.Add(Tokenize(TokenType.BangChar, src.GetSlice())); break;
                case '@': tokens.Add(Tokenize(TokenType.AtSiginChar, src.GetSlice())); break;
                case ',': tokens.Add(Tokenize(TokenType.CommaChar, src.GetSlice())); break;
                case '.': tokens.Add(Tokenize(TokenType.DotChar, src.GetSlice())); break;
                
                case '<': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.LessEqualsOperator, src.GetSlice())
                    : Tokenize(TokenType.LeftAngleChar, src.GetSlice())); break;
                
                case '>': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.GreatEqualsOperator, src.GetSlice())
                    : Tokenize(TokenType.RightAngleChar, src.GetSlice())); break;
                
                case '+': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.AddAssigin, src.GetSlice())
                    : Tokenize(TokenType.CrossChar, src.GetSlice())); break;
                
                case '-': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.SubAssigin, src.GetSlice())
                    : Tokenize(TokenType.MinusChar, src.GetSlice())); break;
                
                case '*': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.MulAssigin, src.GetSlice())
                    : Tokenize(TokenType.StarChar, src.GetSlice())); break;
                
                case '/': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.DivAssigin, src.GetSlice())
                    : Tokenize(TokenType.SlashChar, src.GetSlice())); break;
                
                case '%': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.RestAssigin, src.GetSlice())
                    : Tokenize(TokenType.PercentChar, src.GetSlice())); break;
                
                case '=': tokens.Add(src.NextIs('=')
                    ? Tokenize(TokenType.EqualOperator, src.GetSlice())
                    : Tokenize(TokenType.EqualsChar, src.GetSlice())); break;
                
                default:
                {
                    // Build number token
                    if (char.IsDigit(c))
                    {
                        var numBase = 10;
                        var isFloating = false;
                    
                        if (c == '0' && !src.IsEof()) // verify different bases
                        {
                            if (char.ToLower(src.Peek()) == 'x')
                            {
                                numBase = 16;
                                src.Next();
                            }
                            if (char.ToLower(src.Peek()) == 'b')
                            {
                                numBase = 2;
                                src.Next();
                            }
                        }

                        while (!src.IsEof())
                        {
                            char cc = src.Peek();
                            if (cc == '.')
                            {
                                if (numBase != 10 || isFloating) break;

                                isFloating = true;
                                _ = src.Next();
                                continue;
                            }

                            if (numBase == 10 && !char.IsDigit(src.Peek())) break;
                            if (numBase == 16 && !char.IsAsciiHexDigit(src.Peek())) break;
                            if (numBase == 2  && src.Peek() is not ('0' or '1')) break;
                            
                            src.Next();
                        }

                        tokens.Add(Tokenize(
                            isFloating
                                ? TokenType.FloatingNumberLiteral
                                : TokenType.IntegerNumberLiteral,
                            src.GetSlice()));

                    }

                    // Build identifier token
                    else if (c.IsValidOnIdentifierStarter())
                    {
                        while (src.Peek().IsValidOnIdentifier()) src.Next();

                        var value = src.GetSlice();

                        tokens.Add(_keyword2TokenMap.TryGetValue(value.Item1.ToString(), out var type)
                            ? Tokenize(type, value)
                            : Tokenize(TokenType.Identifier, value));
                    }

                    // Build string tokens
                    else if (c == '"')
                    {
                        tokens.Add(Tokenize(TokenType.DoubleQuotes, src.GetSlice()));

                        while (!src.IsEof())
                        {
                            // test escape
                            if (src.Peek() == '\\')
                            {
                                if (src.Dirty) tokens.Add(Tokenize(TokenType.StringLiteral, src.GetSlice()));
                            
                                if (src.IsEof()) break;
                                char escapeCode = src.Next();
                            
                                // interpolation
                                if (escapeCode == '{') throw new NotImplementedException();
                            
                                // hexadecimal
                                else if (escapeCode == 'x')
                                {
                                    while (src.Peek() != '\\') src.Next(); 
                                    tokens.Add(Tokenize(TokenType.CharacterLiteral, src.GetSlice()));
                                }
                                else tokens.Add(Tokenize(TokenType.CharacterLiteral, src.GetSlice()));
                            }

                            // test end of string
                            else if (src.NextIs('"'))
                            {
                                if (src.Dirty) tokens.Add(Tokenize(TokenType.StringLiteral, src.GetSlice()));
                                tokens.Add(Tokenize(TokenType.DoubleQuotes, src.GetSlice()));
                                break;
                            }
                            else src.Next();
                        }
                    }

                    // Igonore comments
                    else if (c == '#')
                    {
                        if (src.NextStringIs("###"))
                        {
                            while (!src.NextStringIs("###")) src.Next();
                            src.Discard();
                        }
                        else
                        {
                            while (src.Peek() != '\n') src.Next();
                            src.Discard();
                        }
                    }

                    // unrecognized character
                    else
                    {
                        throw new NotImplementedException();
                        // FIXME implement error handlers
                        // try { throw new UnrecognizedCharacterException(c, i); }
                        // catch (SyntaxException e) { currentSrc.ThrowError(e); }
                    }

                    break;
                }
            }
        }
        
        tokens.Add(Tokenize(TokenType.EofChar, src.GetSlice()));

        VerifyEndOfStatements(tokens);

    }

    private void VerifyEndOfStatements(List<Token> tokensList)
    {
        var indexesToRemove = new List<int>();

        for (int index = 0; index < tokensList.Count; index++)
        {
            var currToken = tokensList[index];

            // Se precisa remover quebras de linha à esquerda
            if ((lineJunctions["justLeft"].Contains(currToken.type)
                 || lineJunctions["bothSides"].Contains(currToken.type)) && index > 0)
            {
                for (int i = index - 1; i >= 0; i--)
                {
                    if (tokensList[i].type == TokenType.LineFeedChar)
                    {
                        indexesToRemove.Add(i);
                    }
                    else break;
                }
            }

            // Se precisa remover quebras de linha à direita
            if (lineJunctions["justRight"].Contains(currToken.type)
                || lineJunctions["bothSides"].Contains(currToken.type))
            {
                for (int i = index + 1; i < tokensList.Count; i++)
                {
                    if (tokensList[i].type == TokenType.LineFeedChar)
                    {
                        indexesToRemove.Add(i);
                    }
                    else break;
                }
            }
        }

        // Remove os marcados, de trás pra frente para não quebrar os índices
        foreach (var i in indexesToRemove.Distinct().OrderByDescending(x => x))
            tokensList.RemoveAt(i);

        // 🔎 Passo extra: eliminar duplicados consecutivos de LineFeedChar
        for (int i = tokensList.Count - 2; i >= 0; i--)
        {
            if (tokensList[i].type == TokenType.LineFeedChar &&
                tokensList[i + 1].type == TokenType.LineFeedChar)
            {
                tokensList.RemoveAt(i + 1);
            }
        }
    }

    
    private class SourceWindow(string source)
    {
        private string source = source;
        
        private uint line = 0;
        private uint startColumn = 0;
        private uint endColumn = 0;
        
        private int index_start = 0;
        private int index_end = 0;

        public bool Dirty => index_end != index_start;

        private char Advance()
        {
            if (IsEof()) return (char)0x1A;
            
            var c = source[index_end++];
            endColumn++;

            switch (c)
            {
                case '\n': NextLine(); break;
            }
            
            return c;
        }
        
        private void NextLine()
        {
            line++;
            startColumn = 0;
            endColumn = 0;
            _ = CloseBounds();
        }

        public bool IsEof() => index_end >= source.Length;
        public char Peek()
        {
            if (IsEof()) return (char)0x1A;
            return source[index_end];
        }
        public char Next() => Advance();

        public bool NextIs(char c)
        {
            if (Peek() != c) return false;
            _ = Advance();
            return true;

        }
        public bool NextIfTest(Func<char, bool> test)
        {
            if (test(Peek())) return false;
            _ = Advance();
            return true;
        }

        public bool NextStringIs(string s)
        {
            if (index_end + s.Length >= source.Length) return false;
            var res = source[index_end .. (index_end + s.Length)] == s;
            if (!res) return false;

            for (var i = 0; i < s.Length; i++) Advance();
            return true;
        }
        
        public void Discard() => _ = CloseBounds();
        public (uint, uint, uint) CloseBounds()
        {
            var ret = (line, startColumn, endColumn);
            startColumn = endColumn;
            index_start = index_end;
            return ret;
        }
        public (ReadOnlyMemory<char>, uint, uint, uint) GetSlice()
        {
            var str = source.AsMemory(index_start, index_end - index_start);
            var bounds = CloseBounds();
            return (str, bounds.Item1, bounds.Item2, bounds.Item3); 
        }
        
        public override string ToString()
            => $"({line+1}:{startColumn}-{endColumn})\n{source[0..index_end]}{{{{!index here!}}}}{source[index_end..]}";
    }
}
